rust-runtime/aws-smithy-http-server/src/request/request_id.rs

```diff
@@ -150,3 +183,84 @@ impl<Protocol> IntoResponse<Protocol> for MissingServerRequestId {
         internal_server_error()
     }
 }
+
+struct ResponsePackage {
+    request_id: ServerRequestId,
+    header_key: Option<HeaderName>,
+}
+
+pin_project_lite::pin_project! {
+    pub struct ServerRequestIdResponseFuture<Fut> {
+        response_package: Option<ResponsePackage>,
+        #[pin]
+        fut: Fut,
+    }
+}
+
+impl<Fut> Future for ServerRequestIdResponseFuture<Fut>
+where
+    Fut: TryFuture<Ok = Response<crate::body::BoxBody>>,
+{
+    type Output = Result<Fut::Ok, Fut::Error>;
+
+    fn poll(self: std::pin::Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+        let fut = this.fut;
+        let response_package = this.response_package;
+        fut.try_poll(cx)
+            .map_ok(|mut res| {
+                if let Some(response_package) = response_package {
+                    if let Ok(value) = HeaderValue::from_str(&response_package.request_id.id.to_string()) {
```

Comment:
Can `HeaderValue::from_str(&response_package.request_id.id.to_string())` ever fail?

What is the format of `HeaderValue` in relation to UUID v4?

I'd take this conversion out and place it in a `pub crate ServerRequestId::to_header` method and, if it can never fail, then make said method infallible by the use of `.expect` inside.